pkg/services/authz/rbac/resolver.go

```diff
@@ -0,0 +1,95 @@
+package rbac
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/grafana/authlib/types"
+	"github.com/grafana/grafana/pkg/registry/apis/iam/legacy"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+)
+
+type ScopeResolverFunc func(id string) (string, error)
+
+func (s *Service) newTeamNameResolver(ctx context.Context, ns types.NamespaceInfo) ScopeResolverFunc {
+	teamIDs, ok := s.teamIDCache.Get(ctx, teamIDsCacheKey(ns.Value))
+	if !ok {
+		// TODO singleflight?
+		teams, err := s.identityStore.ListTeams(ctx, ns, legacy.ListTeamQuery{})
```

Comment:
looking through this -- i think this is a solid approach. the other thing i was thinking of when we talked was that you can use unified storage to search based on a certain set of ids, but if we are listing them all in order to populate the cache-- this makes more sense, particularly at the # of teams an instance will have

Reply:
if this becomes an issue, another option could be to populate the cache at startup in the background